### PR TITLE
Store hostname of workerutil Record processor on job dequeue

### PR DESCRIPTION
**Pull request by Strum355**
This PR originally came out of a desire to know what gcloud VM was processing a particular code-intel autoindex job. Previously, if issues occurred relating to a particular host, we would have to go through logs for each host to track down which one was the offending one. This issue is compounded when trying to do aggregated analysis over a longer period of time. As we extract job processing out into separate services (such as through @efritz's [work](https://github.com/sourcegraph/sourcegraph/issues/21756), this data point may become increasingly useful.

A new column is added to the tables of all processes that build upon dbworker/workerutil to track the hostname of the service processing a particular record. The hostname is overridable by setting the `HOSTNAME` environment variable.
Affected tables: `changeset_jobs`, `changesets`, `cm_action_jobs`, `cm_trigger_jobs`, `external_service_sync_jobs`, `insights_query_runner_jobs`, `lsif_dependency_indexing_jobs`, `lsif_indexes`, `lsif_uploads`

### CODE REVIEW

**Review by efritz — APPROVED** [review]
Looks very good!

To confirm, this also won't cause any issues with outdated executors (do we have to line up the deployments to stop the dequeue from failing, or will an empty field for the hostname be alright - it looks to be so from what I can tell)?

**Comment by Strum355** [comment]
for executors this presents no problem. The default for the column is '', and when passed down from the executor-queue, itll be as an empty string anyways.

**Comment by coury-clark** [comment]
Just to understand something - is the intention behind something like this purely for analysis / debugging / reporting purposes, or is this something that application logic is going to utilize to make decisions?

What I'm wondering is what changes for us by storing this field in the table instead of just reporting it in the logs? Somehow the hostname is still going to have to get reported (presumably in logs), so it isn't clear to me what this changes for debugging purposes. 

My guess is that using logger contexts would also address this same issue in a consistent way, which would allow us to use automation for monitoring / alerting / reporting / analysis on these jobs. I've had pretty good success with similar things in the past, where the worker manager would register a bunch of metadata (hostname, job context like table, id, etc) to the logger context and we would extract that in our logging system.

Any thoughts on that? I could also be completely missing some important context here!!

**Comment by Strum355** [comment]
> Just to understand something - is the intention behind something like this purely for analysis / debugging / reporting purposes, or is this something that application logic is going to utilize to make decisions?
> 
> What I'm wondering is what changes for us by storing this field in the table instead of just reporting it in the logs? Somehow the hostname is still going to have to get reported (presumably in logs), so it isn't clear to me what this changes for debugging purposes.
> 
> My guess is that using logger contexts would also address this same issue in a consistent way, which would allow us to use automation for monitoring / alerting / reporting / analysis on these jobs. I've had pretty good success with similar things in the past, where the worker manager would register a bunch of metadata (hostname, job context like table, id, etc) to the logger context and we would extract that in our logging system.
> 
> Any thoughts on that? I could also be completely missing some important context here!!

Completely valid points to make :slightly_smiling_face: There is currently no intention for it to inform application logic (but completely welcome to do so!). There are a few main points I have to provide some context behind this.

1. From the code-intel side, the plan is currently to surface the hostname in the site-admin pages for autoindex jobs (example [here](https://sourcegraph.com/site-admin/code-intelligence/indexes/TFNJRkluZGV4OjgxNzEyNQ==)). This will allow us to more rapidly narrow the search scope without having to search through logs for the specific job logs. That page is most frequently used as the first point of debugging by us (in my opinion its much more direct and faster to process as a first stop than going straight to logs)
2. On the topic of logging systems, the code-intel autoindex infrastructure currently resides in a completely different GCP project (for security reasons), and as such as a less refined logging system. The logs contain a lot of the host operating system/microvm os system logs as well, which adds a lot of noise to the debugging process. These logs would be a next step along the path of debugging issues for us, after we've narrowed down where we look with the info from site-admin n co. The logging system there could definitely use some work, but it would require a larger time investment on multiple fronts that we currently cant provide :slightly_frowning_face: 

If any of this needs more context, lemme know :slightly_smiling_face: @coury-clark

**Comment by camdencheek** [comment]
Anything in particular you're looking for in a review from search-product, or just a notify/stamp?

**Comment by Strum355** [comment]
@camdencheek mostly a stamp, given this affects multiple teams :slightly_smiling_face:

**Review by camdencheek — APPROVED** [review]

File: migrations/frontend/1528395834_dbworkers_worker_hostname.up.sql [hunk 1]
```diff
@@ -0,0 +1,13 @@
+BEGIN;
+
+ALTER TABLE changeset_jobs ADD COLUMN worker_hostname text NOT NULL DEFAULT '';
```
Comment (eseliger) [inline migrations/frontend/1528395834_dbworkers_worker_hostname.up.sql:3]: Shouldn't the field be nullable? When the record is created, no worker is assigned yet, no?

Comment (Strum355) [reply migrations/frontend/1528395834_dbworkers_worker_hostname.up.sql:3]: We decided for NOT NULL and `''` denoting an empty/unset value as this value will always be `!= ''` if it was/is being processed due to the worker setting the `workerutils.Options.WorkerHostname` to a non `''` value if not explicitly provided. Makes it slightly nicer to use on the Go side over *string of sql.NullString. 
Additionally, even if we decided to make it nullable, NULL vs `''` doesnt add any additional information.

File: enterprise/internal/codemonitors/background/workers.go [hunk 2]
```diff
@@ -173,7 +173,6 @@ func (r *queryRunner) Handle(ctx context.Context, workerStore dbworkerstore.Stor
 	err = s.LogSearch(ctx, newQuery, numResults, record.RecordID())
 	if err != nil {
 		return fmt.Errorf("LogSearch: %w", err)
-
```
Comment (unknwon) [inline enterprise/internal/codemonitors/background/workers.go:176]: 👍

File: migrations/frontend/1528395834_dbworkers_worker_hostname.down.sql [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+
+ALTER TABLE changeset_jobs DROP COLUMN worker_hostname;
+ALTER TABLE changesets DROP COLUMN worker_hostname;
+ALTER TABLE cm_action_jobs DROP COLUMN worker_hostname;
+ALTER TABLE cm_trigger_jobs DROP COLUMN worker_hostname;
+ALTER TABLE external_service_sync_jobs DROP COLUMN worker_hostname;
+ALTER TABLE insights_query_runner_jobs DROP COLUMN worker_hostname;
+ALTER TABLE lsif_dependency_indexing_jobs DROP COLUMN worker_hostname;
+ALTER TABLE lsif_indexes DROP COLUMN worker_hostname;
+ALTER TABLE lsif_uploads DROP COLUMN worker_hostname;
```
Comment (unknwon) [inline migrations/frontend/1528395834_dbworkers_worker_hostname.down.sql:11]: nit:
```suggestion
ALTER TABLE changeset_jobs DROP COLUMN IF EXISTS worker_hostname;
ALTER TABLE changesets DROP COLUMN IF EXISTS worker_hostname;
ALTER TABLE cm_action_jobs DROP COLUMN IF EXISTS worker_hostname;
ALTER TABLE cm_trigger_jobs DROP COLUMN IF EXISTS worker_hostname;
ALTER TABLE external_service_sync_jobs DROP COLUMN IF EXISTS worker_hostname;
ALTER TABLE insights_query_runner_jobs DROP COLUMN IF EXISTS worker_hostname;
ALTER TABLE lsif_dependency_indexing_jobs DROP COLUMN IF EXISTS worker_hostname;
ALTER TABLE lsif_indexes DROP COLUMN IF EXISTS worker_hostname;
ALTER TABLE lsif_uploads DROP COLUMN IF EXISTS worker_hostname;
```

File: migrations/frontend/1528395834_dbworkers_worker_hostname.up.sql (hunk 1 above)
Comment (unknwon) [reply migrations/frontend/1528395834_dbworkers_worker_hostname.up.sql:3]: I also find using `''` makes code much nicer looking particularly for Go, unless NULL has special meaning for the value itself (e.g. `deleted_at`).

File: migrations/frontend/1528395834_dbworkers_worker_hostname.up.sql [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+ALTER TABLE changesets ADD COLUMN worker_hostname text NOT NULL DEFAULT '';
+ALTER TABLE cm_action_jobs ADD COLUMN worker_hostname text NOT NULL DEFAULT '';
+ALTER TABLE cm_trigger_jobs ADD COLUMN worker_hostname text NOT NULL DEFAULT '';
+ALTER TABLE external_service_sync_jobs ADD COLUMN worker_hostname text NOT NULL DEFAULT '';
+ALTER TABLE insights_query_runner_jobs ADD COLUMN worker_hostname text NOT NULL DEFAULT '';
+ALTER TABLE lsif_dependency_indexing_jobs ADD COLUMN worker_hostname text NOT NULL DEFAULT '';
+ALTER TABLE lsif_indexes ADD COLUMN worker_hostname text NOT NULL DEFAULT '';
+ALTER TABLE lsif_uploads ADD COLUMN worker_hostname text NOT NULL DEFAULT '';
```
Comment (unknwon) [inline migrations/frontend/1528395834_dbworkers_worker_hostname.up.sql:11]: nit:
```suggestion
ALTER TABLE changesets ADD COLUMN IF NOT EXISTS worker_hostname text NOT NULL DEFAULT '';
ALTER TABLE cm_action_jobs ADD COLUMN IF NOT EXISTS worker_hostname text NOT NULL DEFAULT '';
ALTER TABLE cm_trigger_jobs ADD COLUMN IF NOT EXISTS worker_hostname text NOT NULL DEFAULT '';
ALTER TABLE external_service_sync_jobs ADD COLUMN IF NOT EXISTS worker_hostname text NOT NULL DEFAULT '';
ALTER TABLE insights_query_runner_jobs ADD COLUMN IF NOT EXISTS worker_hostname text NOT NULL DEFAULT '';
ALTER TABLE lsif_dependency_indexing_jobs ADD COLUMN IF NOT EXISTS worker_hostname text NOT NULL DEFAULT '';
ALTER TABLE lsif_indexes ADD COLUMN IF NOT EXISTS worker_hostname text NOT NULL DEFAULT '';
ALTER TABLE lsif_uploads ADD COLUMN IF NOT EXISTS worker_hostname text NOT NULL DEFAULT '';
```

**Review by unknwon — APPROVED** [review]

**Pull request merged**